File: backend/app/schemas/marks.py
"""Pydantic schemas for marks and exams endpoints."""
from typing import List, Optional
from pydantic import BaseModel


class ExamOut(BaseModel):
    id: str
    name: str
    term: str
    academic_year: str
    start_date: str
    end_date: str
    status: str   # upcoming | ongoing | completed

    model_config = {"from_attributes": True}


class MarkRecordOut(BaseModel):
    id: str
    exam_subject_id: str
    student_id: str
    marks_obtained: float
    grade: str
    remarks: Optional[str] = None
    # Joined fields
    subject_name: Optional[str] = None
    subject_code: Optional[str] = None
    max_marks: Optional[float] = None
    passing_marks: Optional[float] = None
    exam_name: Optional[str] = None
    exam_date: Optional[str] = None

    model_config = {"from_attributes": True}


class EnterMarkRequest(BaseModel):
    exam_subject_id: str
    student_id: str
    marks_obtained: float
    remarks: Optional[str] = None


class MarkReportOut(BaseModel):
    """Assembled report card for a student in one exam."""
    student_id: str
    student_name: str
    exam: ExamOut
    marks: List[MarkRecordOut]
    total_obtained: float
    total_max: float
    percentage: float
    overall_grade: str

File: src/services/marks.ts
import { api } from './api';

export interface Exam {
  id: string;
  name: string;
  term: string;
  academic_year: string;
  start_date: string;
  end_date: string;
  status: 'upcoming' | 'ongoing' | 'completed';
}

export interface MarkRecord {
  id: string;
  exam_subject_id: string;
  student_id: string;
  marks_obtained: number;
  grade: string;
  remarks?: string;
  subject_name?: string;
  subject_code?: string;
  max_marks?: number;
  passing_marks?: number;
  exam_name?: string;
  exam_date?: string;
}

export async function getExams(): Promise<Exam[]> {
  return api.get<Exam[]>('/marks/exams');
}

export async function getStudentMarks(studentId: string, examId?: string): Promise<MarkRecord[]> {
  const query = examId ? `?exam_id=${examId}` : '';
  return api.get<MarkRecord[]>(`/marks/student/${studentId}${query}`);
}

export async function enterMark(examSubjectId: string, studentId: string, marksObtained: number, remarks?: string): Promise<MarkRecord> {
  return api.post<MarkRecord>('/marks/enter', {
    exam_subject_id: examSubjectId,
    student_id: studentId,
    marks_obtained: marksObtained,
    remarks,
  });
}
